src/checkfiles/checkfiles.py

```diff
@@ -206,7 +229,77 @@ def bam_pysam_check(file_path, file_format='bam'):
+    # reference_file looks like this: /reference-files/TSTFI36924773/
+    accession = reference_file.split('/')[-2]
+    reference_file_path = os.path.join(
+        SUPPORTING_FILES_FOLDER, accession + '.fasta')
+    if not os.path.exists(reference_file_path):
+        # a download link looks like this: https://api.sandbox.igvf.org/reference-files/TSTFI36924773/@@download/TSTFI36924773.fasta.gz
+        download_link = f'https://api.sandbox.igvf.org/reference-files/{accession}/@@download/{accession}.fasta.gz'
+        # download the file
+        session = requests.Session()
+        session.auth = portal_auth
+        response = session.get(download_link)
+        if response.status_code == 200:
+            try:
+                # Decompress gzip and write .fasta file
+                with gzip.open(BytesIO(response.content), 'rt') as gz_file:
+                    with open(reference_file_path, 'w') as f:
+                        f.write(gz_file.read())
+            except Exception as e:
+                raise RuntimeError(
+                    f'Failed to decompress or save the reference file: {e}')
+        else:
+            raise RuntimeError(
+                f'Failed to download file: {download_link} (status {response.status_code})')
+    return reference_file_path
+
+
+def cram_pysam_check(file_path, reference_file_path):
+    error = {}
+    try:
+        pysam.quickcheck(file_path)
+        # First command: samtools view
+        view_cmd = [
+            'samtools', 'view',
+            '-h',
+            '-T', reference_file_path,
+            file_path
+        ]
+        # Second command: samtools stats -
+        stats_cmd = [
+            'samtools', 'stats', '-'
+        ]
+
+        with subprocess.Popen(view_cmd, stdout=subprocess.PIPE, stderr=subprocess.PIPE) as p1:
+            with subprocess.Popen(stats_cmd, stdin=p1.stdout, stdout=subprocess.PIPE, stderr=subprocess.PIPE, text=True) as p2:
+                p1.stdout.close()  # Let p1 receive SIGPIPE if p2 exits
+                _, err_view = p1.communicate()  # Capture stderr from view
+                if err_view:
+                    error = {'cram_error': f'samtools view error: {err_view}'}
+                    return error
+                output, err_stats = p2.communicate()
+                if err_stats:
+                    error = {'cram_error': f'samtools stats error: {err_stats}'}
+                    return error
+                if 'SN\tis sorted:\t0' in output:
+                    error = {'cram_error': f'the cram file is not sorted'}
+                    return error
+                with pysam.AlignmentFile(file_path, 'rc', reference_filename=reference_file_path) as cram:
+
+                    count = cram.count(until_eof=True)
+                    logger.info(f'the number of reads: {count}')
+                    info = {'read_count': count}
+                    return info
+
+    except pysam.utils.SamtoolsError as e:
```

Comment:
wonder if we should also be catching `subprocess.CalledProcessError` just so that some creatively malformed cram would not crash the checkfiles.

Reply:
Sounds good. Will add it.